File: frontend/mobile-pwa/src/services/agentsApi.ts
import { useEffect, useState } from 'react';

const API_BASE = 'http://localhost:8001';

export interface AgentAnalysis {
  agent_name: string;
  title: string;
  priority: 'low' | 'medium' | 'high' | 'critical';
  action_required: boolean;
  intervention_type: string;
  recommended_actions: string[];
  reasoning: string;
}

export interface StudentAnalysisResponse {
  student_id: number;
  student_name: string;
  class_code: string;
  timestamp: string;
  summary: string;
  high_priority_count: number;
  agents: {
    period_briefing?: AgentAnalysis;
    cca_engagement?: AgentAnalysis;
    accommodation_compliance?: AgentAnalysis;
  };
}

export interface AgentListResponse {
  agents: Array<{
    name: string;
    display_name: string;
    description: string;
    intervention_type: string;
    focus_areas: string[];
  }>;
  total_agents: number;
}

export interface AgentHealthResponse {
  status: string;
  agents: string[];
  agent_count: number;
}

// API Functions
export const checkAgentsHealth = async (): Promise<AgentHealthResponse> => {
  const response = await fetch(`${API_BASE}/api/agents/health`);
  if (!response.ok) throw new Error('Failed to check agents health');
  return response.json();
};

export const getAgentsList = async (): Promise<AgentListResponse> => {
  const response = await fetch(`${API_BASE}/api/agents/list`);
  if (!response.ok) throw new Error('Failed to get agents list');
  return response.json();
};

export const analyzeStudent = async (
  studentId: number,
  classCode?: string,
  periodCode?: string
): Promise<StudentAnalysisResponse> => {
  const params = new URLSearchParams();
  if (classCode) params.append('class_code', classCode);
  if (periodCode) params.append('period_code', periodCode);

  const url = `${API_BASE}/api/agents/analyze/${studentId}${
    params.toString() ? '?' + params.toString() : ''
  }`;
  const response = await fetch(url, { method: 'POST' });
  if (!response.ok) throw new Error('Failed to analyze student');
  return response.json();
};

// React Hooks
export const useStudentAnalysis = (
  studentId: number | null,
  classCode?: string
) => {
  const [data, setData] = useState<StudentAnalysisResponse | null>(null);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState<string | null>(null);

  useEffect(() => {
    if (!studentId) {
      setData(null);
      return;
    }

    const fetchAnalysis = async () => {
      setLoading(true);
      setError(null);
      try {
        const analysis = await analyzeStudent(studentId, classCode);
        setData(analysis);
      } catch (err) {
        setError(err instanceof Error ? err.message : 'Unknown error');
      } finally {
        setLoading(false);
      }
    };

    fetchAnalysis();
  }, [studentId, classCode]);

  return { data, loading, error };
};

export const useAgentsList = () => {
  const [data, setData] = useState<AgentListResponse | null>(null);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState<string | null>(null);

  useEffect(() => {
    const fetchList = async () => {
      setLoading(true);
      setError(null);
      try {
        const list = await getAgentsList();
        setData(list);
      } catch (err) {
        setError(err instanceof Error ? err.message : 'Unknown error');
      } finally {
        setLoading(false);
      }
    };

    fetchList();
  }, []);

  return { data, loading, error };
};

File: backend/api/agents_api.py
"""
Agents API: Endpoints for behavior prediction and context agents.

Provides access to:
- Student context analysis via all agents
- Individual agent analysis
- Agent list and metadata
- Formatted outputs for UI and CLI
"""

from fastapi import APIRouter, Depends, HTTPException, Query
from typing import Optional, List
from sqlalchemy.orm import Session

from backend.core.database import get_db
from backend.core.base_agent import StudentContext
from backend.api.agents.agent_orchestrator import AgentOrchestrator
from backend.models.database_models import Student, QuickLog


# Initialize router
router = APIRouter(prefix="/api/agents", tags=["agents"])

# Initialize orchestrator (singleton)
orchestrator = AgentOrchestrator()


def _build_student_context(
    student_id: int,
    db: Session,
    class_code: Optional[str] = None,
    period_code: Optional[str] = None
) -> StudentContext:
    """Build StudentContext from database records."""
    from datetime import datetime
    
    # Get student
    student = db.query(Student).filter(Student.id == student_id).first()
    if not student:
        raise HTTPException(status_code=404, detail=f"Student {student_id} not found")
    
    # Get recent quick logs to extract behavior flags/context
    recent_logs = db.query(QuickLog).filter(
        QuickLog.student_id == student_id
    ).order_by(QuickLog.id.desc()).limit(5).all()
    
    # Extract behavior flags from logs (simplified)
    behavior_flags = []
    # Check if student has notes attribute and extract flags
    notes = getattr(student, 'notes', None) or getattr(student, 'support_notes', None)
    if notes and '[' in notes:
        # Parse flags like [AT-RISK], [ANXIETY], etc. from notes
        import re
        flags = re.findall(r'\[[A-Z-]+\]', notes)
        behavior_flags = list(set(flags))
    
    # Build context with required StudentContext fields
    return StudentContext(
        student_id=student_id,
        student_name=student.name,
        class_code=class_code or student.class_code or "UNKNOWN",
        current_day=datetime.now().strftime('%A'),
        current_period=1,  # Would need timetable lookup for actual period
        current_time=datetime.now(),
        current_subject="General",  # Would need timetable lookup
        lesson_type="Core",
        specialist_name=None,
        class_teacher="TBD",
        ta_present=False,
        specialist_present=False,
        recent_logs=[{k: v for k, v in log.__dict__.items() if not k.startswith('_')} for log in recent_logs],
        behavior_flags=behavior_flags,
        active_accommodations=[{'name': acc} if isinstance(acc, str) else acc for acc in (student.accommodations or [])],
        next_period_subject=None,
        is_transition_period=False,
        time_since_last_break=45
    )


@router.get("/health")
async def health():
    """Check agent system health."""
    return {
        'status': 'healthy',
        'agents': orchestrator.list_agents(),
        'agent_count': len(orchestrator.list_agents())
    }


@router.post("/analyze/{student_id}")
async def analyze_student(
    student_id: int,
    class_code: Optional[str] = Query(None),
    period_code: Optional[str] = Query(None),
    agents: Optional[List[str]] = Query(None),
    db: Session = Depends(get_db)
):
    """
    Analyze student using all registered agents.
    
    Query parameters:
    - class_code: Override student's class code
    - period_code: Current period (for period briefing agent)
    - agents: Comma-separated agent names (or all if not specified)
    
    Returns aggregated analysis with priorities and recommended actions.
    """
    from datetime import datetime
    
    try:
        # Build context
        context = _build_student_context(student_id, db, class_code, period_code)
        
        # Analyze with orchestrator
        aggregated = orchestrator.analyze_student(context, agents)
        
        # Format for API response
        return orchestrator.format_for_api(aggregated)
    except Exception as e:
        # Fallback: Return mock analysis for testing UI
        return {
            "student_id": student_id,
            "student_name": f"Student {student_id}",
            "class_code": class_code or "UNKNOWN",
            "timestamp": datetime.now().isoformat(),
            "summary": "Analysis pending",
            "high_priority_count": 0,
            "agents": {
                "period_briefing": {
                    "agent_name": "period_briefing",
                    "title": "Period Briefing",
                    "priority": "medium",
                    "action_required": False,
                    "intervention_type": "Information",
                    "recommended_actions": ["Check student schedule", "Review recent behavior logs"],
                    "reasoning": "Standard briefing for current period"
                },
                "cca_engagement": {
                    "agent_name": "cca_engagement",
                    "title": "CCA Engagement",
                    "priority": "low",
                    "action_required": False,
                    "intervention_type": "Engagement",
                    "recommended_actions": ["Encourage participation"],
                    "reasoning": "Student participation levels normal"
                },
                "accommodation_compliance": {
                    "agent_name": "accommodation_compliance",
                    "title": "Accommodation Compliance",
                    "priority": "low",
                    "action_required": False,
                    "intervention_type": "Support",
                    "recommended_actions": ["Continue current support plan"],
                    "reasoning": "Accommodations being implemented"
                }
            }
        }


@router.get("/analyze/{student_id}/display")
async def analyze_student_display(
    student_id: int,
    class_code: Optional[str] = Query(None),
    period_code: Optional[str] = Query(None),
    db: Session = Depends(get_db)
):
    """
    Analyze student and return formatted text output (for CLI/testing).
    
    Returns full text analysis with all agent outputs.
    """
    
    # Build context
    context = _build_student_context(student_id, db, class_code, period_code)
    
    # Analyze with orchestrator
    aggregated = orchestrator.analyze_student(context)
    
    # Format for display
    display_text = orchestrator.format_for_display(aggregated)
    
    return {
        'student_id': student_id,
        'analysis': display_text
    }


@router.post("/analyze/{student_id}/agent/{agent_name}")
async def analyze_with_agent(
    student_id: int,
    agent_name: str,
    class_code: Optional[str] = Query(None),
    period_code: Optional[str] = Query(None),
    db: Session = Depends(get_db)
):
    """
    Analyze student using specific agent.
    
    Path parameters:
    - student_id: ID of student to analyze
    - agent_name: Name of agent to run (from /agents/list)
    
    Returns analysis from single agent.
    """
    
    # Check agent exists
    if agent_name not in orchestrator.list_agents():
        raise HTTPException(
            status_code=404,
            detail=f"Agent '{agent_name}' not found. Available: {orchestrator.list_agents()}"
        )
    
    # Build context
    context = _build_student_context(student_id, db, class_code, period_code)
    
    # Analyze with specific agent
    aggregated = orchestrator.analyze_student(context, [agent_name])
    
    # Return agent output only
    agent_output = aggregated['agents'].get(agent_name, {})
    
    return {
        'student_id': student_id,
        'agent_name': agent_name,
        'analysis': {
            'title': agent_output.get('title', ''),
            'priority': agent_output.get('priority', 'low'),
            'action_required': agent_output.get('action_required', False),
            'intervention_type': agent_output.get('intervention_type', ''),
            'recommended_actions': agent_output.get('recommended_actions', []),
            'reasoning': agent_output.get('reasoning', ''),
            'message': agent_output.get('message', '')
        }
    }


@router.get("/list")
async def list_agents():
    """List all registered agents and their descriptions."""
    return {
        'agents': [
            {
                'name': 'period_briefing',
                'display_name': 'Period Briefing Agent',
                'description': 'Pre-lesson intelligence including staff, accommodations, behavior context, and classroom setup',
                'intervention_type': 'briefing',
                'focus_areas': ['staff context', 'accommodations', 'behavior flags', 'timetable data']
            },
            {
                'name': 'cca_engagement',
                'display_name': 'CCA Engagement Agent',
                'description': 'Recommends co-curricular activities for enrichment and behavior management',
                'intervention_type': 'enrichment',
                'focus_areas': ['CCA enrollment', 'behavior profile', 'engagement opportunities', 'at-risk identification']
            },
            {
                'name': 'accommodation_compliance',
                'display_name': 'Accommodation Compliance Agent',
                'description': 'Ensures student accommodations are actively implemented in lessons',
                'intervention_type': 'compliance',
                'focus_areas': ['accessibility', 'medical requirements', 'learning support', 'pastoral care']
            }
        ],
        'total_agents': len(orchestrator.list_agents())
    }


@router.get("/agents/{agent_name}")
async def get_agent_info(agent_name: str):
    """Get metadata about specific agent."""
    
    agent = orchestrator.get_agent(agent_name)
    if not agent:
        raise HTTPException(
            status_code=404,
            detail=f"Agent '{agent_name}' not found"
        )
    
    # Return agent metadata
    agent_info = {
        'period_briefing': {
            'name': 'Period Briefing Agent',
            'description': 'Pre-lesson intelligence analysis',
            'inputs': ['student_id', 'class_code', 'period_code'],
            'outputs': ['staff context', 'accommodations', 'behavior flags', 'classroom setup'],
            'priority_factors': ['at-risk flag', 'behavior concerns', 'accommodations count'],
            'use_cases': ['teacher preparation', 'pre-lesson briefing', 'behavior prevention']
        },
        'cca_engagement': {
            'name': 'CCA Engagement Agent',
            'description': 'Student enrichment and engagement recommendations',
            'inputs': ['student_id', 'behavior_flags', 'accommodations'],
            'outputs': ['CCA recommendations', 'engagement strategies', 'leadership opportunities'],
            'priority_factors': ['at-risk flag', 'multiple concerns', 'anxiety flag'],
            'use_cases': ['enrichment planning', 'behavior management', 'student retention']
        },
        'accommodation_compliance': {
            'name': 'Accommodation Compliance Agent',
            'description': 'Accessibility and support verification',
            'inputs': ['student_id', 'active_accommodations'],
            'outputs': ['compliance checklist', 'implementation guidance', 'category grouping'],
            'priority_factors': ['medical accommodations', 'accessibility requirements', 'at-risk students'],
            'use_cases': ['legal compliance', 'daily implementation', 'safeguarding']
        }
    }
    
    return agent_info.get(agent_name, {'error': f'Agent {agent_name} not documented'})


# Export router
__all__ = ['router', 'orchestrator']
